src/EcommerceStatsBundle/Service/StatsTableCsvExportService.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+namespace ChameleonSystem\EcommerceStatsBundle\Service;
+
+use ChameleonSystem\EcommerceStatsBundle\DataModel\StatsGroupDataModel;
+use ChameleonSystem\EcommerceStatsBundle\DataModel\StatsTableDataModel;
+use ChameleonSystem\EcommerceStatsBundle\Interfaces\StatsTableCsvExportServiceInterface;
+use Symfony\Component\Translation\TranslatorInterface;
+use TCMSLocal;
+
+class StatsTableCsvExportService implements StatsTableCsvExportServiceInterface
+{
+    private const TRANSLATION_TOTAL = 'chameleon_system_ecommerce_stats.total';
+    private const TRANSLATION_DELTA = 'chameleon_system_ecommerce_stats.delta';
+
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    /**
+     * @var TCMSLocal
+     */
+    private $local;
+
+    public function __construct(TranslatorInterface $translator)
+    {
+        $this->translator = $translator;
+        $this->local = TCMSLocal::GetActive();
+    }
+
+    public function getCSVData(StatsTableDataModel $statsTable): array
+    {
+        $data = [
+            $this->getHeaderRow($statsTable),
+        ];
+
+        foreach ($statsTable->getBlocks() as $block) {
+            $this->exportBlockCSV($data, $statsTable, $block, 1);
+            $data[] = [];
+        }
+
+        return $data;
+    }
+
+    private function getHeaderRow(StatsTableDataModel $statsTable): array
+    {
+        $row = array_fill(0, $statsTable->getMaxGroupCount(), '');
+
+        foreach ($statsTable->getColumnNames() as $name) {
+            $row[] = $name;
+            if ($statsTable->isShowDiffColumn()) {
+                $row[] = $this->translator->trans(self::TRANSLATION_DELTA);
+            }
+        }
+
+        return $row;
+    }
+
+    protected function exportBlockCSV(
+        array &$data,
```

Comment:
Does the & still make a difference?

Reply:
Yes, the method updates the array and passes it down recursively. The alternative to it would be a whole lot of array_merge calls.

Reply:
But passing it down (and change there) should also work without & (?).

Reply:
Not really, since in PHP arrays are not passed by reference, but by value: Each method call with array contents actually copies the array. It would work, if we would be using a `Collection` of some sort, since that would be an instance of a class, which would be passed by reference by default.

As long as it's a primitive (which an array is in PHP), passing by reference needs to be 'enabled' explicitly